src/lib.mo

```diff
@@ -1,10 +1,1100 @@
-/// ...
+/// A module which implements auction functionality for various trading pairs against "trusted" fungible token
 ///
 /// Copyright: 2023-2024 MR Research AG
-/// Main author: 
-/// Contributors: 
+/// Main author: Andy Gura
+/// Contributors: Timo Hanke
 
+import Array "mo:base/Array";
+import AssocList "mo:base/AssocList";
+import Float "mo:base/Float";
+import Int "mo:base/Int";
+import Iter "mo:base/Iter";
+import List "mo:base/List";
+import Nat "mo:base/Nat";
+import Nat64 "mo:base/Nat64";
+import Option "mo:base/Option";
 import Prim "mo:prim";
+import Principal "mo:base/Principal";
+import R "mo:base/Result";
+import RBTree "mo:base/RBTree";
+
+import Vec "mo:vector";
 
 module {
+
+  public type AssetId = Nat;
+  public type OrderId = Nat;
+
+  public type StableDataV1 = {
+    counters : (sessions : Nat, orders : Nat);
+    assets : Vec.Vector<StableAssetInfo>;
+    users : RBTree.Tree<Principal, UserInfo>;
+    history : List.List<PriceHistoryItem>;
+    stats : {
+      usersAmount : Nat;
+      accountsAmount : Nat;
+      assets : Vec.Vector<{ bidsAmount : Nat; totalBidVolume : Nat; asksAmount : Nat; totalAskVolume : Nat; lastProcessingInstructions : Nat }>;
+    };
+  };
+
+  public func defaultStableDataV1() : StableDataV1 = {
+    counters = (0, 0);
+    assets = Vec.new();
+    users = #leaf;
+    history = null;
+    stats = {
+      usersAmount = 0;
+      accountsAmount = 0;
+      assets = Vec.new();
+    };
+  };
+
+  public type Account = {
+    // balance of user account
+    var credit : Nat;
+    // user's credit, placed as bid or ask
+    var lockedCredit : Nat;
+  };
+
+  public type PriceHistoryItem = (timestamp : Nat64, sessionNumber : Nat, assetId : AssetId, volume : Nat, price : Float);
+  public type TransactionHistoryItem = (timestamp : Nat64, sessionNumber : Nat, kind : { #ask; #bid }, assetId : AssetId, volume : Nat, price : Float);
+
+  public type Order = {
+    user : Principal;
+    userInfoRef : UserInfo;
+    assetId : AssetId;
+    price : Float;
+    var volume : Nat;
+  };
+
+  public type SharedOrder = {
+    assetId : AssetId;
+    price : Float;
+    volume : Nat;
+  };
+
+  public type UserInfo = {
+    var currentAsks : AssocList.AssocList<OrderId, Order>;
+    var currentBids : AssocList.AssocList<OrderId, Order>;
+    var credits : AssocList.AssocList<AssetId, Account>;
+    var history : List.List<TransactionHistoryItem>;
+  };
+
+  public type StableAssetInfo = {
+    asks : AssocList.AssocList<OrderId, Order>;
+    bids : AssocList.AssocList<OrderId, Order>;
+    lastSwapRate : Float;
```

Comment:
Need to use a different name. "Swap rate" has a meaning in financial markets (related to interest rate swaps). Just `lastRate` should be fine.

Reply:
Renamed